addons/mail/static/src/rtc/voice_analyser.js

```diff
@@ -0,0 +1,103 @@
+ * @extends {Component<Props, Env>}
+ */
+export class VoiceAnalyser extends Component {
+    static props = ["attachment"];
+    static template = "mail.VoiceAnalyser";
+
+    duration;
+    dataArray;
+    startTime;
+    analyser;
+    pausedElapsed = 0;
+
+    setup() {
+        /** @type {import("@mail/core/sound_effects_service").SoundEffects} */
+        this.soundEffectsService = useService("mail.sound_effects");
+        this.audioRef = useRef("audio");
+        this.canvasRef = useRef("canvas");
+        this.duration = this.props.attachment.voiceDuration;
+        this.state = useState({
+            playing: false,
+            paused: false,
+            intervalId: false,
+            visualTime: this.generateTime(Math.floor(this.duration / 1000)),
+        });
+
+        onMounted(async () => {
+            this.primaryColor = getComputedStyle(this.audioRef.el).getPropertyValue("--primary");
+            this.ctx = this.canvasRef.el.getContext("2d");
+            this.ctx.fillStyle = this.primaryColor;
+            this.fillBackground();
+        });
+    }
+
+    fillBackground() {
+        for (var i = 0; i < this.canvasRef.el.width; i += 6) {
+            this.ctx.beginPath();
+            this.ctx.arc(i, this.canvasRef.el.height / 2, 2, 0, 2 * Math.PI);
+            this.ctx.fill();
+        }
+    }
+
+    renderFrame(t) {
+        if (!this.canvasRef.el) {
+            return;
+        }
+        if (this.state.paused && this.startTime) {
+            this.pausedElapsed = t - this.startTime;
+            return;
+        }
+        if (!this.startTime) {
+            this.startTime = t;
+        } else if (this.pausedElapsed) {
+            this.startTime = t - this.pausedElapsed;
+            this.pausedElapsed = 0;
+        }
+        const elapsed = t - this.startTime;
+        this.analyser.getByteTimeDomainData(this.dataArray);
+
+        const min = Math.min(...this.dataArray);
+        const max = Math.max(...this.dataArray);
+        const barHeight = ((max - min) / 128) * this.canvasRef.el.height * 2;
+        const x = (elapsed / this.duration) * this.canvasRef.el.width;
+        this.ctx.fillRect(x, (this.canvasRef.el.height - barHeight) / 2, 1, barHeight);
+
+        if (elapsed < this.duration) {
+            requestAnimationFrame(this.renderFrame.bind(this));
+        } else {
+            this.state.playing = false;
+            this.startTime = false;
+            clearInterval(this.state.intervalId);
+            this.soundEffectsService.activeAnalyser = null;
+            this.state.visualTime = this.generateTime(Math.floor(this.duration / 1000));
+            this.ctx.clearRect(0, 0, this.canvasRef.el.width, this.canvasRef.el.height);
+            this.fillBackground();
+        }
+    }
+
+    SetIntervalId() {
+        let totalSeconds = 0;
+        if (this.pausedElapsed) {
+            totalSeconds = Math.floor(this.pausedElapsed / 1000);
+        }
+        this.state.intervalId = setInterval(() => {
```

Comment:
It looks like this interval is never stopped

Reply:
It's closed on stopping voice analyser on `onClickAnalyser` and at the end of playing voice message on `renderFrame`

Reply:
I think that if you close (destroy the component when it's playing) without stopping (so without clicking for `onClickAnalyser()`) it will not get stopped by the `renderFrame()` because `renderFrame()` will return early.

Reply:
Tu as raison. I have to stop it on `onWillUnmount` too.